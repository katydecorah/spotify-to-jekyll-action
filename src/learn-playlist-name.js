import { exportVariable } from "@actions/core";

export default function learnPlaylistName() {
  const today = new Date();
  const month = process.env.MONTH || today.getMonth();
  const year = process.env.YEAR || today.getFullYear();
  const season = {
    2: "Winter",
    5: "Spring",
    8: "Summer",
    11: "Fall",
  };
  const name = `${month == 2 ? `${year - 1}/${year}` : year} ${season[month]}`;
  exportVariable("playlist", name);
  return name;
}
