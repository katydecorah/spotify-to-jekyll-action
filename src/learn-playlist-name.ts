import { getInput } from "@actions/core";
import * as github from "@actions/github";

export default function learnPlaylistName(): string {
  const payload = github.context.payload.inputs;
  if (payload && payload["playlist-name"]) {
    return payload["playlist-name"];
  }

  const today = new Date();
  const month = process.env.MONTH
    ? parseInt(process.env.MONTH)
    : today.getMonth();
  const year = process.env.YEAR
    ? parseInt(process.env.YEAR)
    : today.getFullYear();
  const [marchEnd, juneEnd, septemberEnd, decemberEnd] = validateSeasonNames();
  const seasons = {
    2: marchEnd,
    5: juneEnd,
    8: septemberEnd,
    11: decemberEnd,
  };
  const season = seasons[month];
  if (!season)
    throw new Error(`The current month does not match an end of season month.`);
  return `${month === 2 ? `${year - 1}/${year}` : year} ${season}`;
}

function validateSeasonNames() {
  const seasonNames = getInput("season-names")
    .split(",")
    .map((s) => s.trim());
  if (seasonNames.length !== 4)
    throw new Error(
      `There must be 4 seasons listed in \`season-names\` only found ${seasonNames.length} (\`${seasonNames}\`).`
    );
  return seasonNames;
}
