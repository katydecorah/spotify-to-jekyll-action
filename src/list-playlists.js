import { setFailed } from "@actions/core";
import SpotifyWebApi from "spotify-web-api-node";

export default async function listPlaylists(listName) {
  const spotifyApi = new SpotifyWebApi({
    clientId: process.env.SpotifyClientID,
    clientSecret: process.env.SpotifyClientSecret,
  });
  const username = process.env.SpotifyUser;
  const {
    body: { access_token },
  } = await spotifyApi.clientCredentialsGrant();

  spotifyApi.setAccessToken(access_token);

  const { body } = await spotifyApi.getUserPlaylists(username);
  const findPlaylist = body.items.find((list) => list.name === listName);
  if (!findPlaylist) {
    setFailed(`Could not find playlist "${listName}". Is it private?`);
    return;
  }
  const {
    body: { items },
  } = await spotifyApi.getPlaylistTracks(findPlaylist.id);
  return {
    name: findPlaylist.name,
    external_urls: findPlaylist.external_urls,
    images: findPlaylist.images,
    tracks: {
      items,
    },
  };
}
