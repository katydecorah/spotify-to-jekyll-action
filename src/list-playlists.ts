import { getInput } from "@actions/core";
import SpotifyWebApi from "spotify-web-api-node";
import { Playlist } from "./index.js";

export default async function listPlaylists(
  listName: string
): Promise<Playlist> {
  try {
    const spotifyApi = new SpotifyWebApi({
      clientId: process.env.SpotifyClientID,
      clientSecret: process.env.SpotifyClientSecret,
    });
    const username = getInput("spotify-username");
    const {
      body: { access_token },
    } = await spotifyApi.clientCredentialsGrant();

    spotifyApi.setAccessToken(access_token);

    const { body } = await spotifyApi.getUserPlaylists(username);
    const findPlaylist = body.items.find(({ name }) => name === listName);
    if (!findPlaylist) {
      throw new Error(`Could not find playlist "${listName}". Is it private?`);
    }
    const {
      body: { items },
    } = await spotifyApi.getPlaylistTracks(findPlaylist.id);
    if (!items.length) throw new Error("Playlist has no tracks.");

    return formatTracks({
      name: findPlaylist.name,
      external_urls: findPlaylist.external_urls,
      images: findPlaylist.images,
      tracks: items,
    });
  } catch (error) {
    throw new Error(error);
  }
}

export function formatTracks({
  name,
  external_urls,
  images,
  tracks,
}: {
  name: string;
  external_urls: SpotifyApi.ExternalUrlObject;
  images: SpotifyApi.ImageObject[];
  tracks: SpotifyApi.PlaylistTrackObject[];
}): Playlist {
  const largestImage = images.sort(
    (a, b) => (b.width || 0) - (a.width || 0)
  )[0];
  return {
    name,
    formatted_name: formatName(name),
    url: external_urls.spotify,
    tracks: tracks.map(({ track }) => ({
      name: track?.name,
      artist: track?.artists.map(({ name }) => name).join(", "),
      album: track?.album.name,
    })),
    image: largestImage.url,
  };
}

function formatName(name: string): string {
  return name
    .replace(/\s/g, "-")
    .replace(/[^a-zA-Z0-9-]/g, "")
    .replace(/-+/g, "-")
    .toLowerCase();
}
