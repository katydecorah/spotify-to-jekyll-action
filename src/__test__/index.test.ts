import { action } from "..";
import updateMain from "spotify-to-jekyll/src/update-main.js";
import { exportVariable } from "@actions/core";
import * as ListPlaylists from "../list-playlists";
import * as LearnPlaylistName from "../learn-playlist-name";

jest.mock("@actions/core");
jest.mock("spotify-to-jekyll/src/update-main.js");
jest.mock("fs");
jest.mock("spotify-web-api-node", () => {
  return jest.fn().mockImplementation(() => {
    const userPlaylists = require("./fixtures/get-user-playlists.json");
    const playlistTracks = require("./fixtures/get-playlist-tracks.json");
    return {
      clientCredentialsGrant: jest
        .fn()
        .mockImplementation(() =>
          Promise.resolve({ body: { access_token: "token" } })
        ),
      setAccessToken: jest.fn(),
      getUserPlaylists: jest
        .fn()
        .mockImplementation(() => Promise.resolve({ body: userPlaylists })),
      getPlaylistTracks: jest
        .fn()
        .mockImplementation(() =>
          Promise.resolve({ body: { items: playlistTracks } })
        ),
    };
  });
});

describe("action", () => {
  test("works", async () => {
    process.env.MONTH = "11";
    process.env.YEAR = "2021";
    const learnPlaylistNameSpy = jest.spyOn(LearnPlaylistName, "default");
    const listPlaylistsSpy = jest.spyOn(ListPlaylists, "default");
    await action();
    expect(learnPlaylistNameSpy).toHaveReturnedWith("2021 Fall");
    const results = listPlaylistsSpy.mock.results[0].value;
    await expect(results).resolves.toMatchSnapshot();
    expect(exportVariable).toHaveBeenNthCalledWith(1, "playlist", "2021 Fall");
    expect(exportVariable).toHaveBeenNthCalledWith(
      2,
      "PlaylistImageOutput",
      "playlist-2021-fall.png"
    );
    expect(exportVariable).toHaveBeenNthCalledWith(
      3,
      "PlaylistImage",
      "https://mosaic.scdn.co/640/ab67616d0000b27304a205b54b5c1fbae8543efaab67616d0000b27364fa86e376b0279aa49a1832ab67616d0000b273acddc5a09b8cb4730995b60aab67616d0000b273ba738766498bddccae3d319a"
    );
    expect(updateMain.mock.calls[0]).toMatchInlineSnapshot(`
      Array [
        Object {
          "formatted_name": "2021-fall",
          "image": "playlist-2021-fall.png",
          "name": "2021 Fall",
          "tracks": Array [
            Object {
              "album": "one hand on the steering wheel the other sewing a garden",
              "artist": "Ada Lea",
              "name": "can't stop me from dying",
            },
            Object {
              "album": "No Shadow",
              "artist": "Hyd",
              "name": "No Shadow",
            },
            Object {
              "album": "Trip To Japan",
              "artist": "The Shacks",
              "name": "Trip To Japan",
            },
            Object {
              "album": "Keeper",
              "artist": "Hana Vu",
              "name": "Maker",
            },
            Object {
              "album": "Doomin' Sun",
              "artist": "Bachelor, Jay Som, Palehound",
              "name": "Anything at All",
            },
            Object {
              "album": "Would You Mind Please Pulling Me Close?",
              "artist": "Tasha",
              "name": "Would You Mind Please Pulling Me Close?",
            },
            Object {
              "album": "Genesis",
              "artist": "Spencer.",
              "name": "Genesis",
            },
            Object {
              "album": "Old Peel",
              "artist": "Aldous Harding",
              "name": "Old Peel",
            },
            Object {
              "album": "Everybody's Birthday",
              "artist": "Hana Vu",
              "name": "Everybody's Birthday",
            },
            Object {
              "album": "The Baby",
              "artist": "Samia",
              "name": "Big Wheel",
            },
            Object {
              "album": "-io",
              "artist": "Circuit des Yeux",
              "name": "Dogma",
            },
            Object {
              "album": "private LIFE",
              "artist": "Virginia Wing",
              "name": "I'm Holding Out For Something",
            },
            Object {
              "album": "Blue Weekend",
              "artist": "Wolf Alice",
              "name": "Delicious Things",
            },
            Object {
              "album": "Bottle Episode",
              "artist": "Mandy, Indiana",
              "name": "Bottle Episode",
            },
            Object {
              "album": "The Gaping Mouth",
              "artist": "Lowertown",
              "name": "The Gaping Mouth",
            },
            Object {
              "album": "Fantasize Your Ghost",
              "artist": "Ohmme",
              "name": "3 2 4 3",
            },
            Object {
              "album": "You Think It's Like This But Really It's Like This",
              "artist": "Mirah",
              "name": "Of Pressure",
            },
            Object {
              "album": "When the Sun Comes Up",
              "artist": "Greta Morgan",
              "name": "When the Sun Comes Up",
            },
            Object {
              "album": "Ceremony",
              "artist": "Anna von Hausswolff",
              "name": "Mountains Crave",
            },
          ],
          "url": "https://open.spotify.com/playlist/2YnPs9UNBkJpswmsRNwQ1o",
        },
      ]
    `);
  });
});
