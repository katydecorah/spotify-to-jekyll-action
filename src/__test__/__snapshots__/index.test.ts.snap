// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action works 1`] = `
Object {
  "formatted_name": "2021-fall",
  "image": "playlist-2021-fall.png",
  "name": "2021 Fall",
  "tracks": Array [
    Object {
      "album": "one hand on the steering wheel the other sewing a garden",
      "artist": "Ada Lea",
      "name": "can't stop me from dying",
    },
    Object {
      "album": "No Shadow",
      "artist": "Hyd",
      "name": "No Shadow",
    },
    Object {
      "album": "Trip To Japan",
      "artist": "The Shacks",
      "name": "Trip To Japan",
    },
    Object {
      "album": "Keeper",
      "artist": "Hana Vu",
      "name": "Maker",
    },
    Object {
      "album": "Doomin' Sun",
      "artist": "Bachelor, Jay Som, Palehound",
      "name": "Anything at All",
    },
    Object {
      "album": "Would You Mind Please Pulling Me Close?",
      "artist": "Tasha",
      "name": "Would You Mind Please Pulling Me Close?",
    },
    Object {
      "album": "Genesis",
      "artist": "Spencer.",
      "name": "Genesis",
    },
    Object {
      "album": "Old Peel",
      "artist": "Aldous Harding",
      "name": "Old Peel",
    },
    Object {
      "album": "Everybody's Birthday",
      "artist": "Hana Vu",
      "name": "Everybody's Birthday",
    },
    Object {
      "album": "The Baby",
      "artist": "Samia",
      "name": "Big Wheel",
    },
    Object {
      "album": "-io",
      "artist": "Circuit des Yeux",
      "name": "Dogma",
    },
    Object {
      "album": "private LIFE",
      "artist": "Virginia Wing",
      "name": "I'm Holding Out For Something",
    },
    Object {
      "album": "Blue Weekend",
      "artist": "Wolf Alice",
      "name": "Delicious Things",
    },
    Object {
      "album": "Bottle Episode",
      "artist": "Mandy, Indiana",
      "name": "Bottle Episode",
    },
    Object {
      "album": "The Gaping Mouth",
      "artist": "Lowertown",
      "name": "The Gaping Mouth",
    },
    Object {
      "album": "Fantasize Your Ghost",
      "artist": "Ohmme",
      "name": "3 2 4 3",
    },
    Object {
      "album": "You Think It's Like This But Really It's Like This",
      "artist": "Mirah",
      "name": "Of Pressure",
    },
    Object {
      "album": "When the Sun Comes Up",
      "artist": "Greta Morgan",
      "name": "When the Sun Comes Up",
    },
    Object {
      "album": "Ceremony",
      "artist": "Anna von Hausswolff",
      "name": "Mountains Crave",
    },
  ],
  "url": "https://open.spotify.com/playlist/2YnPs9UNBkJpswmsRNwQ1o",
}
`;
